### PR TITLE
Added char_is_uppercase.

### PR DESCRIPTION
**Pull request by Lildirt**
Is char_is_uppercase acceptable?

### CODE REVIEW

File: src/main/java/com/laytonsmith/core/functions/StringHandling.java [hunk 1]
```diff
@@ -845,6 +846,184 @@ public Construct exec(Target t, Environment env, Construct... args) throws Cance
...
+			String text = args[0].nval();
+			
+			// Enforce the fact we are only taking the first character here
+			// Do not let the user pass an entire string. Only a single character.
+			if(text.length() > 1) {
+			    throw new CRECastException("Got " + text + " instead of expected character.", t);
+			}
+			
+			char check = text.charAt(0);
+			if(Character.isUpperCase(check)) {
```
Comment (LadyCailin) [inline src/main/java/com/laytonsmith/core/functions/StringHandling.java]: return CBoolean.get(Character.isUpperCase(check)); No need to make this into 6 lines.

File: src/main/java/com/laytonsmith/core/functions/StringHandling.java [hunk 2]
```diff
@@ -845,6 +846,184 @@ public Construct exec(Target t, Environment env, Construct... args) throws Cance
...
+			return null;
+		}
+
+		@Override
+		public CHVersion since() {
+		    return CHVersion.V3_3_2;
+		}
+	}
+
+	@api
```
Comment (LadyCailin) [inline src/main/java/com/laytonsmith/core/functions/StringHandling.java]: These functions are incredibly specific. char_is_uppercase is fine, but what is the use case for adding this to the core? I would prefer to let users write a proc to do this.

**Comment by Lildirt** [comment]
You have a point. PK mentioned this as well and, yeah, it's probably best to let the user handle this.

File: src/main/java/com/laytonsmith/core/functions/StringHandling.java [hunk 3]
```diff
@@ -845,6 +846,63 @@ public Construct exec(Target t, Environment env, Construct... args) throws Cance
 	}
 
 	@api
+	public static class char_is_uppercase extends AbstractFunction {
```
Comment (LadyCailin) [inline src/main/java/com/laytonsmith/core/functions/StringHandling.java]: Add unit tests for this to make sure it works. Particularly, add a unit test that checks for an empty string, a string with more than 1 letter, a lowercase char and an uppercase char.

File: src/main/java/com/laytonsmith/core/functions/StringHandling.java [hunk 4]
```diff
@@ -845,6 +846,63 @@ public Construct exec(Target t, Environment env, Construct... args) throws Cance
...
+		@Override
+		public Boolean runAsync() {
+			return null;
+		}
+
+		@Override
+		public CHVersion since() {
+		    return CHVersion.V3_3_2;
+		}
+	}
```
Comment (LadyCailin) [inline src/main/java/com/laytonsmith/core/functions/StringHandling.java]: Examples would be good too.

File: src/main/java/com/laytonsmith/core/functions/StringHandling.java [hunk 5]
```diff
@@ -865,18 +865,29 @@ public Construct exec(Target t, Environment environment, Construct... args) thro
 				    args[0].typeof() + " was found.", t);
 			}
 			String text = args[0].nval();
-			
 			// Enforce the fact we are only taking the first character here
 			// Do not let the user pass an entire string. Only a single character.
 			if(text.length() > 1) {
-			    throw new CRECastException("Got " + text + " instead of expected character.", t);
+			    throw new CRECastException("Got \"" + text + "\" instead of expected character.", t);
```
Comment (LadyCailin) [inline src/main/java/com/laytonsmith/core/functions/StringHandling.java]: A CREFormatException would be more appropriate here, as well as the exception below. CRECastException is reserved for when, say, an int is expected, but a string is given. In this case, a string would still be given, it would just have the wrong format. See: http://wiki.sk89q.com/wiki/CommandHelper/Staged/Exceptions#FormatException

File: src/main/java/com/laytonsmith/core/functions/StringHandling.java [hunk 6]
```diff
@@ -845,6 +846,82 @@ public Construct exec(Target t, Environment env, Construct... args) throws Cance
...
+		public Construct exec(Target t, Environment environment, Construct... args) throws CancelCommandException, ConfigRuntimeException {
+			if(!(args[0].nval() instanceof String)) {
+			    throw new CREFormatException(this.getName() + " expects a string as first argument, but type " +
+				    args[0].typeof() + " was found.", t);
+			}
+			String text = args[0].nval();
+			// Enforce the fact we are only taking the first character here
+			// Do not let the user pass an entire string. Only a single character.
+			if(text.length() > 1) {
+			    throw new CRECastException("Got \"" + text + "\" instead of expected character.", t);
```
Comment (LadyCailin) [inline src/main/java/com/laytonsmith/core/functions/StringHandling.java]: A CREFormatException would be more appropriate here, as well as the exception below. CRECastException is reserved for when, say, an int is expected, but a string is given. In this case, a string would still be given, it would just have the wrong format. See: http://wiki.sk89q.com/wiki/CommandHelper/Staged/Exceptions#FormatException


The exception above should be a CastException though.

File: src/test/java/com/laytonsmith/core/functions/StringHandlingTest.java [hunk 7]
```diff
@@ -243,6 +245,52 @@ public void testSplitWithLimit2() throws Exception {
 		assertEquals("a", SRun("char_from_unicode(unicode_from_char('a'))", null));
 	}
 
+	@Test
+	public void testCharIsUppercase() throws Exception {
+		assertEquals("true", SRun("char_is_uppercase('M')", null));
+		assertEquals("false", SRun("char_is_uppercase('m')", null));
```
Comment (LadyCailin) [inline src/test/java/com/laytonsmith/core/functions/StringHandlingTest.java]: You forgot to add a test for char_is_uppercase('').

File: src/test/java/com/laytonsmith/core/functions/StringHandlingTest.java [hunk 8]
```diff
@@ -243,6 +245,52 @@ public void testSplitWithLimit2() throws Exception {
...
+		}
+		catch (ConfigCompileException | CREFormatException | CRECastException e) {
+			//pass
+		}
+		
+		try {
+			SRun("char_is_uppercase(dyn('stuff'))", null);
+			fail("Expected char_is_uppercase(dyn('stuff')) to throw an exception.");
+		}
+		catch (ConfigCompileException | CRECastException e) {
```
Comment (LadyCailin) [inline src/test/java/com/laytonsmith/core/functions/StringHandlingTest.java]: We don't expect a ConfigCompileException here

File: src/test/java/com/laytonsmith/core/functions/StringHandlingTest.java [hunk 9]
```diff
@@ -243,6 +245,52 @@ public void testSplitWithLimit2() throws Exception {
...
+			//pass
+		}
+		
+		try {
+			SRun("char_is_uppercase(1)", null);
+			fail("Expected char_is_uppercase(1) to throw an exception.");
+		}
+		catch (ConfigCompileException | CREFormatException | CRECastException e) {
```
Comment (LadyCailin) [inline src/test/java/com/laytonsmith/core/functions/StringHandlingTest.java]: We should only expect one type of exception here

File: src/test/java/com/laytonsmith/core/functions/StringHandlingTest.java [hunk 10]
```diff
@@ -243,6 +245,52 @@ public void testSplitWithLimit2() throws Exception {
...
+		}
+		catch (ConfigCompileException | CRECastException e) {
+			//pass
+		}
+
+		try {
+			SRun("char_is_uppercase('!')", null);
+			fail("Expected char_is_uppercase('!') to throw an exception.");
+		}
+		catch (ConfigCompileException | CRECastException e) {
```
Comment (LadyCailin) [inline src/test/java/com/laytonsmith/core/functions/StringHandlingTest.java]: We only expect a ConfigCompileException here.

File: src/test/java/com/laytonsmith/core/functions/StringHandlingTest.java [hunk 11]
```diff
@@ -243,6 +245,52 @@ public void testSplitWithLimit2() throws Exception {
...
+			//pass
+		}
+
+		try {
+			SRun("char_is_uppercase('1')", null);
+			fail("Expected char_is_uppercase('1') to throw an exception.");
+		}
+		catch (ConfigCompileException | CRECastException e) {
```
Comment (LadyCailin) [inline src/test/java/com/laytonsmith/core/functions/StringHandlingTest.java]: We only expect a ConfigCompileException here.

File: src/test/java/com/laytonsmith/core/functions/StringHandlingTest.java [hunk 12]
```diff
@@ -243,6 +245,52 @@ public void testSplitWithLimit2() throws Exception {
...
+	@Test
+	public void testCharIsUppercase() throws Exception {
+		assertEquals("true", SRun("char_is_uppercase('M')", null));
+		assertEquals("false", SRun("char_is_uppercase('m')", null));
+
+		try {
+			SRun("char_is_uppercase('magic')", null);
+			fail("Expected char_is_uppercase('magic') to throw an exception.");
+		}
+		catch (ConfigCompileException | CRECastException e) {
```
Comment (LadyCailin) [inline src/test/java/com/laytonsmith/core/functions/StringHandlingTest.java]: We only expect a ConfigCompileException here.

File: src/main/java/com/laytonsmith/core/functions/StringHandling.java [hunk 13]
```diff
@@ -845,6 +846,82 @@ public Construct exec(Target t, Environment env, Construct... args) throws Cance
...
+		public CHVersion since() {
+		    return CHVersion.V3_3_2;
+		}
+		
+		
+		@Override
+		public ExampleScript[] examples() throws ConfigCompileException {
+			return new ExampleScript[]{
+						new ExampleScript("", "char_is_uppercase('a')"),
+						new ExampleScript("", "char_is_uppercase('D')"),};
```
Comment (LadyCailin) [inline src/main/java/com/laytonsmith/core/functions/StringHandling.java]: In the first string, put "Basic usage"

File: src/main/java/com/laytonsmith/core/functions/StringHandling.java [hunk 14]
```diff
@@ -845,6 +846,82 @@ public Construct exec(Target t, Environment env, Construct... args) throws Cance
...
+		}
+
+		@Override
+		public Integer[] numArgs() {
+			return new Integer[]{1};
+		}
+
+		@Override
+		public String docs() {
+			return "boolean {char} Determines if the character provided is uppercase or not.";
```
Comment (LadyCailin) [inline src/main/java/com/laytonsmith/core/functions/StringHandling.java]: Change this to

"boolean {char} Determines if the character provided is uppercase or not. The string must be exactly 1 character long and a letter, otherwise a FormatException is thrown."
